llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13795,13 +13798,22 @@ static SDValue combineVWADDWSelect(SDNode *N, SelectionDAG &DAG) {
 
   // False value of MergeOp should be all zeros
   SDValue Z = MergeOp->getOperand(2);
-  if (Z.getOpcode() != ISD::INSERT_SUBVECTOR)
-    return SDValue();
-  if (!ISD::isBuildVectorAllZeros(Z.getOperand(1).getNode()))
-    return SDValue();
-  if (!isNullOrNullSplat(Z.getOperand(0)) && !Z.getOperand(0).isUndef())
+
+  // Scalable vector
+  if (MergeOpc == ISD::VSELECT &&
+      !ISD::isConstantSplatVectorAllZeros(Z.getNode()))
```

Comment:
Does `ISD::isConstantSplatVectorAllZeros` also work for the fixed length case? I presume the `ISD::INSERT_SUBVECTOR` is for converting a fixed vector to scalable. Would it be easier to read if we first "unwrapped" the false value if needed, and then do the `isConstantSplatVectorAllZeros` check?

Reply:
`ISD::isConstantSplatVectorAllZeros` does not work for the fixed length case. Opcode need to be `ISD::BUILD_VECTOR`.

```cpp
bool ISD::isConstantSplatVectorAllZeros(const SDNode *N, bool BuildVectorOnly) {
  // Look through a bit convert.
  while (N->getOpcode() == ISD::BITCAST)
    N = N->getOperand(0).getNode();

  if (!BuildVectorOnly && N->getOpcode() == ISD::SPLAT_VECTOR) {
    APInt SplatVal;
    return isConstantSplatVector(N, SplatVal) && SplatVal.isZero();
  }

  if (N->getOpcode() != ISD::BUILD_VECTOR) return false;
  ...
}
```

Reply:
> I presume the ISD::INSERT_SUBVECTOR is for converting a fixed vector to scalable.
Yes, you are right.

Reply:
The BUILD_VECTOR should be removed before the last round of DAG combine. What does the fixed vector code look like in the last DAG combine?

Reply:
But for the fixed length vector case aren't we checking for a build_vector anyway? 

I think we could unify the paths if we first tried to unwrap Z like this:
(insert_subvector null/undef, build_vector ...) -> (build_vector ...)

Reply:
> The BUILD_VECTOR should be removed before the last round of DAG combine. What does the fixed vector code look like in the last DAG combine?

It will be `(extract_subvector (RISCVISD::VMV_V_X_VL...`

Reply:
> But for the fixed length vector case aren't we checking for a build_vector anyway?
> 
> I think we could unify the paths if we first tried to unwrap Z like this: (insert_subvector null/undef, build_vector ...) -> (build_vector ...)

Okay, thanks for the advice. I will try this way.